Fixed not updating modified date when doing bulk update in Redirect (#9859)

polakluk opened the pull request:
Pull Request for Issue #9859 & #9873.
#### Summary of Changes

Added code to update "modified_date" when doing bulk update for Destination URL do Link model. Also, the "Destination URL" functionality was moved into a separate method in model in order to fix #9873 in clear way.
#### Testing Instructions #9859

_(Copied from the original issue)_
##### Steps to reproduce the issue

The redirect component has a fairly hidden feature to mass update existing links - its at the bottom of the list called Destination URL.
Select some existing links and click on Destination URL
Enter a new Destination URL and a comment
And then update links button to process
##### Expected result

Destination url is changed
comment is added
Last Updated Date is modified
#### Testing Instructions #9873 (ADDED LATER)

_(Copied from the original issue)_
##### Steps to reproduce the issue
1. Go to "Redirect" component in back-end and select some URLs (some of them should be disabled)
2. Click on link "Destination URL" (below table)
3. Enter some URL
4. Leave "Comment" field empty
5. Click on "Update links"
##### Expected result
- Selected rows will have new URL
- Their state (Enabled/disabled) will remain the same
- Their comment will remain the same

Commit:
Fixed joomla/joomla-cms#9859 (redirect update modified date)

brianteeman commented:
I have tested this item :white_check_mark: successfully on 535436eba1aaed4b9daec575f4289f2d58afd919<br /><br />Thanks - I should have worked that fix out myself ;)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9864">issues.joomla.org/joomla-cms/9864</a>.</sub>

Commit:
Update link.php

Deleted whitespace in line 178 making Travis go crazy

andrepereiradasilva commented:
I have tested this item :white_check_mark: successfully on b059fb1cd8437697a6647aafb97989aa5c42202d<br /><br />works as described.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9864">issues.joomla.org/joomla-cms/9864</a>.</sub>

MATsxm commented:
I have tested this item :white_check_mark: successfully on b059fb1cd8437697a6647aafb97989aa5c42202d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9864">issues.joomla.org/joomla-cms/9864</a>.</sub>

brianteeman commented:
@polakluk Once Travis is happy this can be made RTC

 178 | ERROR | Whitespace found at end of line
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9864">issues.joomla.org/joomla-cms/9864</a>.</sub>

andrepereiradasilva commented:
travis is already happy

brianteeman commented:
RTC - thanks - travis took a little while to update the issue tracker ;)
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9864">issues.joomla.org/joomla-cms/9864</a>.</sub>

Commit:
Update link.php

Created a separate method for duplication of URLs, because the inner logic is different from activating a URL.The activate method is left the same as before this PR as it worked correctly (enabling a record does not change modified date of the recor)

Commit:
Update links.php

Added a separate method to handle duplication of URLs

Commit:
Update default_addform.php

Changed task name

polakluk commented:
I added code that also fixes #9873 to this PR in order to keep it in one place

andrepereiradasilva commented:
Now you added new methods ... and i think that means it will go only for 3.6.

I would prefer to keep them separated, this one was already tested and RTC... but, of course, it's your choice.

polakluk commented:
My intention was to keep merging process simple. Since these changes occur in the same file, it is easier to apply one PR to code base than having to merge 2 separate PR and keeping in mind their order (the second part of this PR reverts the change made in the first commit). I am sorry for committing the code after the original PR was already tested. If it causes more trouble than it fixes, then I can create a separate PR for this.

I didnt know that once new methods are added, the patch will be schedule for the next minor version. If that is indeed true, then I can change the comments

andrepereiradasilva commented:
> I didnt know that once new methods are added, the patch will be schedule for the next minor version. If that is indeed true, then I can change the comments

I'm not sure that it's how it works? is it? @brianteeman

Commit:
Update link.php

fixed coding style error for Travis

teccrow commented:
@polakluk I have tested. The expected result is okay, but the create date and the last modify date are always GMT time. Check this, if you change the server time zone. You can this compare with the create date an modify date from article. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9864">issues.joomla.org/joomla-cms/9864</a>.</sub>

teccrow commented:
@polakluk I have the items trashed and the trash is empty. But the redirect function is still running! The database table is also empty. Is this a cookie problem?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9864">issues.joomla.org/joomla-cms/9864</a>.</sub>

polakluk commented:
@teccrow 
I checked the first one and you're right. The fields always show GTM time, because that's how the time is stored (correctly) in DB. However, this seems to be a behavior that is consistent across all parts of Joomla admin (check Articles for example), so this may be intentional. I would like to hear second opinion on this.

As for your second comment. I am still testing it, because my local Joomla installation is wierdly acting out right now.

andrepereiradasilva commented:
@rdeutz i think this one was merged by mistake, the latest big changes were not tested.

let's hope this does not introduces any issue. will test tomorrow.

rdeutz commented:
ups, ok, would be great if you can test, we can revert it if test fail, thanks @andrepereiradasilva

andrepereiradasilva commented:
Tested successfully. So it seems it was a good merge :)

teccrow commented:
@polakluk Correct, the is stored GMT in DB. But, if the Server Time (global Configuration) set for example on 'Berlin' than all Times show in Berlin Time. The form field 'created_date' and 'modify_date' are to be changed.
For example:

        <field 
            name="created_date"
            type="calendar"
            label=""
            description=""
            format="%Y-%m-%d %H:%M:%S"
            size="20"
            readonly="true"
            filter="user_utc" />

Important are the type and the filter!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9864">issues.joomla.org/joomla-cms/9864</a>.</sub>
